### PR TITLE
Improve allocation deciders

### PR DESCRIPTION
**Pull request by idegtiarenko**
This change:
* reduces code duplication
* reduces unnecessary Decision.Multi allocations
* consolidates logging
* Decision.NO is returned with its explanation

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecidersTests.java [hunk 1]
```diff
@@ -20,255 +20,135 @@
...
 
 import static org.hamcrest.Matchers.equalTo;
 
 public class AllocationDecidersTests extends ESTestCase {
 
-    public void testDebugMode() {
-        verifyDebugMode(RoutingAllocation.DebugMode.ON, Matchers.hasSize(1));
-    }
+    public void testCallAllDecidersWhenThereAreNoNoDecisions() {
+        var expectedDecision = switch (randomIntBetween(0, 2)) {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecidersTests.java:40]: likewise here, although I'm not too worried about this one.

**Review by DaveCTurner — COMMENTED** [review]
I like it. I have a slight concern about adding some allocations here, this is still sometimes a bit of a hot path, but I think it's ok.

However I'd rather have this as two separate changes, one that reworks the tests as you propose (but leaves the production implementation alone) and the other which makes the production changes. That way we can be more confident we're not changing behaviour in any important way.

**Comment by idegtiarenko** [comment]
Extracted the test changes here: https://github.com/elastic/elasticsearch/pull/93420
Also fixed test related comments there

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecidersTests.java [hunk 2]
```diff
@@ -53,18 +54,18 @@ public void testCheckAllDecidersBeforeReturningThrottle() {
         var allDecisions = generateDecisions(Decision.THROTTLE, () -> Decision.YES);
         var debugMode = randomFrom(RoutingAllocation.DebugMode.values());
         var expectedDecision = switch (debugMode) {
-            case OFF, EXCLUDE_YES_DECISIONS -> new Decision.Multi().add(Decision.THROTTLE);
+            case OFF -> Decision.THROTTLE;
+            case EXCLUDE_YES_DECISIONS -> new Decision.Multi().add(Decision.THROTTLE);
```
Comment (idegtiarenko) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecidersTests.java:58]: Same here

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecidersTests.java [hunk 3]
```diff
@@ -53,18 +54,18 @@ public void testCheckAllDecidersBeforeReturningThrottle() {
...
     }
 
     public void testExitsAfterFirstNoDecision() {
-        var expectedDecision = Decision.NO;
-        var terminatingDecision = randomFrom(Decision.NO, Decision.single(Decision.Type.NO, "no with label", "explanation"));
-        var allDecisions = generateDecisions(terminatingDecision, () -> randomFrom(Decision.YES, Decision.THROTTLE));
-        var expectedCalls = allDecisions.indexOf(terminatingDecision) + 1;
+        var expectedDecision = randomFrom(Decision.NO, Decision.single(Decision.Type.NO, "no with label", "explanation"));
+        var allDecisions = generateDecisions(expectedDecision, () -> randomFrom(Decision.YES, Decision.THROTTLE));
+        var expectedCalls = allDecisions.indexOf(expectedDecision) + 1;
```
Comment (idegtiarenko) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecidersTests.java:68]: Updated code bypasses actual decision with explanation (if any) to the caller code.
Please let me know if this should be reverted.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDeciders.java [hunk 4]
```diff
@@ -34,279 +37,202 @@ public class AllocationDeciders {
...
+        BiFunction<String, Decision, String> logMessageCreator
+    ) {
+        if (debugMode == RoutingAllocation.DebugMode.OFF) {
+            var result = Decision.YES;
+            for (AllocationDecider decider : deciders) {
+                var decision = deciderAction.apply(decider);
+                if (decision.type() == Decision.Type.NO) {
+                    if (logger.isTraceEnabled()) {
+                        logger.trace(() -> logMessageCreator.apply(decider.getClass().getSimpleName(), decision));
+                    }
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDeciders.java:208]: I wonder if there is a value in logging throttling decisions as well as they could prevent allocations

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecidersTests.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecidersTests.java:68]: should be fine 👍

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDeciders.java (hunk 4 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDeciders.java:208]: I'm ok with this (for now at least)

**Review by DaveCTurner — APPROVED** [review]
LGTM nice simplification

**Comment by idegtiarenko** [comment]
@elasticsearchmachine please run elasticsearch-ci/part-1

**Comment by idegtiarenko** [comment]
@elasticsearchmachine please run elasticsearch-ci/part-1

**Pull request merged**